Fix some template definitions for clang 12

Bensk1 opened the pull request:
Fixes #2370
 
Apparently, clang12 (default when installing llvm via brew on macOS) causes some trouble with template definitions for release builds. This PR should fix the issue. It also removes the unnecessary template of `JoinHash::calculate_radix_bits()`.

(Compile time difference on nemea was negligible.)

Bouncner commented:
I am working (for quite a while already to be honest) on the radix stuff and the template argument makes sense to estimate the hash map size. I am fine with removing it for now, but I am curious: is there any other reason to remove than "it' just not needed"?

Bensk1 commented:
> I am working (for quite a while already to be honest) on the radix stuff and the template argument makes sense to estimate the hash map size. I am fine with removing it for now, but I am curious: is there any other reason to remove than "it' just not needed"?

No, not really. I had to move its definition from .cpp to .hpp in #2360 because gcc10 complained. When discussion the clang12 template issues with Markus, he saw that `calculate_radix_bits` does never use the template. Because of that, I removed the template and moved the implementation back to the .cpp file. Do I understand correctly that you are planning to use the template parameter?

Bouncner commented:
Yes, I use it in a branch that is blocked by some other branch. I don't know if I will ever open a PR for them. So no reason to not remove the template now. I was just curious if there was a more "technical reason".

mweisgut commented:
I cannot run `hyriseTest` (`RelWithDebInfo` build) successfully (using brew llvm 12):
```
[ RUN      ] ExpressionEvaluatorToValuesTest.TernaryOrLiterals
dyld: lazy symbol binding failed: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: <path>/bRelWithDebInfo/./hyriseTest
  Expected in: flat namespace

dyld: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: <path>/bRelWithDebInfo/./hyriseTest
  Expected in: flat namespace

[1]    15931 abort      ./hyriseTest
```

@Bensk1 I'm not sure which commit I successfully build and run in the afternoon. Based on the timestamps, I assume it was 6028d66 or 453cd70.

Bensk1 commented:
> Edit: [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) doesn't result in the above error and runs successfully, [453cd70](https://github.com/hyrise/hyrise/commit/453cd703f97184234fdcd36950aa27b779877123) does result in the above error.

This is really strange because 6028d66 does not change anything except removing two comments.

mweisgut commented:
> > Edit: [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) doesn't result in the above error and runs successfully, [453cd70](https://github.com/hyrise/hyrise/commit/453cd703f97184234fdcd36950aa27b779877123) does result in the above error.
> 
> This is really strange because [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) does not change anything except removing two comments.

But 6028d66 contains the `ExpressionEvaluator` template definitions whereas 453cd70 does not. So with the template definitions, the above error doesn't occur, without the template definitions, the error occurs.

I will try to build it with llvm 12.0.0_1 and get back to you.

Bensk1 commented:
Ah, I misread the commit history on GitHub. It shows newest commits at the bottom, the opposite way of `git log`.

mweisgut commented:
I also tried (multiple times) to build the latest version of this branch using llvm 12.0.0_1 and the `dyld: lazy symbol binding failed` error in `ExpressionEvaluatorToValuesTest.TernaryOrLiterals` still occurred. Using version 6028d66, `hyriseTest` runs without any errors.

mweisgut commented:
#2370

Bensk1 commented:
Should work now. @mweisgut, could you give it another try?

mweisgut commented:
`hyriseTest` successfully runs now in `Release` and `RelWithDebInfo` modes.

julianmenzler commented on src/lib/expression/evaluation/expression_evaluator.cpp:
```diff
@@ -1613,4 +1613,14 @@ std::vector<std::shared_ptr<ExpressionResult<Result>>> ExpressionEvaluator::_pru
   return results;
 }
 
+// We explicitly instantiate these template functions because (at least) clang-12 does not instantiate them for us.
```
Maybe also mention the platform macOS here?

mweisgut replied:
For this, we should verify that the error does not appear on ubuntu with clang 12.

julianmenzler replied:
Does the CI use clang 12?

Bensk1 replied:
> Does the CI use clang 12?

No, it doesn't.

julianmenzler replied:
Good. Let's leave the comment as is and merge this PR.

Bouncner replied:
Any reason not to keep clang9 and use clang12 over clang10 in the CI?

Bensk1 replied:
> Any reason not to keep clang9 and use clang12 over clang10 in the CI?

Not really but we might have to upgrade the Ubuntu version then. Also, with the Mac ARM CI we will have clang12 as part of the CI.
